[2단계 - 자주 가는 음식점] 밍고(성민제) 미션 제출합니다.

MinSungJe opened the pull request:
안녕하세요, 케빈! 너무 오랜만에 뵙습니다!!
개인적으로 난이도가 있기도 했고, 이번 주 테코톡 발표 준비가 겹쳐 많이 미흡하다고 생각되는 코드입니다 😢
그래도 나름대로 E2E 테스트를 이용한 TDD와 컴포넌트 단위 재사용을 고려해보았습니다.
보기 괴로울? 순 있어도 이번 2단계 미션도 잘 부탁드립니다..!!

---

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://minsungje.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

![image](https://github.com/user-attachments/assets/632ded80-3818-4e94-af7f-fb3ef4852b31)


1단계에서 모든 기능을 클래스 없이 구현했기에, **2단계의 기능도 클래스 없이 모두 함수형으로 구현**했습니다.
그 과정에서 컴포넌트를 재사용할 때 오히려 클래스가 가지는 장점을 느낄 수 있었습니다.

클래스의 인스턴스는 자체적으로 값을 보관할 수 있습니다.
그렇기 때문에 인스턴스 별 구분을 위한 ID를 지정하면 이를 쉽게 활용할 수 있습니다.
또한 인스턴스별 메서드를 제공하기 때문에 동작의 책임을 구분하기 쉽습니다.

하지만 저의 경우에는 모든 경우를 함수형으로 구현했습니다.
특히 `RestaurantList`의 경우 모든 음식점 목록과 즐겨찾기한 음식점 목록을 표시하기 위해 총 두 개의 RestaurantList를 사용했는데요.
이 과정에서 개별 RestaurantList의 목록을 업데이트하기 위해선 별도의 id가 필요하기에 **RestaurantList에서 사용하는 모든 함수에 id를 인자로 전달**해야 했습니다.
따라서 다른 컴포넌트(FilterSelect) 등에서 특정 RestaurantList에 접근하기 위해선 불편하게 인자로 전달해야 했습니다.
이를 통해, 특정 구성에 맞춰 비슷한 기능을 가진 컴포넌트를 찍어낼 수 있다는 점에서 함수보다 클래스가 이번 미션에 더 유용한 구현 방법이라는 점을 배울 수 있었습니다.

이외에도 TypeScript를 익히기 위해 UI가 아닌 코드와 이에 대한 단위 테스트 코드를 전부 TS로 변환해 작성했습니다..

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

### applyFilter의 위치

`applyFilter()`는 다음과 같은 역할을 하는 함수입니다.
```js
applyFilter(restaurantListId) {
  // Filter의 값을 가져와서,
  const category, sortingRule = ...

  // Filter를 적용한 List를 계산하고
  const filteredListByBoth = filter(category, sortingRule 이용)

  // 이를 RestaurantList에 적용
  RestaurantList.applyList(restaurantListId, filteredListByBoth);
},
```
이렇게 만든 `applyFilter()`를 이용하면 다른 컴포넌트에서 `applyFilter()`함수를 호출하면 언제든 필터된 음식점 목록을 RestaurantList에 띄울 수 있습니다.

하지만 이 함수를 처음 만들었을 때 위치를 어디에 둬야 할 지 감이 오지 않았습니다.

RestaurantList의 값을 바꾸는 것이기에 이를 RestaurantList에 위치해야하나 싶다가도, 한편으로는 **RestaurantList가 Filter를 알 필요가 있을까?** 싶기도 합니다. 즉 RestaurantList는 주어지는 List대로 목록을 출력하는 책임만 지고, 필터링된 리스트를 제공하는 것은 바깥에서 제공하는 겁니다.

그래서 지금은 필터드롭다운을 만드는 FilterSelect에 해당 함수를 위치해두었습니다. 이 때문에 이 함수를 호출하려면 `FilterSelect.applyFilter()`가 됩니다. 만들고 여러번 사용하면서 이게 맞나...? 싶은 부분이 있었기에 이에 대한 케빈의 의견을 듣고 싶습니다.

### 상태 관리

이번 점심 뭐먹지는 렌더링과의 싸움이었습니다.

- 음식점 목록에서 즐겨찾기를 누르면 아이콘 변화를 위해 음식점 목록을 리렌더링해야 됨
- DetailModal에서 즐겨찾기를 눌렀을 때 바깥 컴포넌트에도 적용되야 하므로 리렌더링
- 삭제하기 버튼을 눌렀을 때도 리렌더링
- 필터를 고를때도 맞춰 목록이 표시되야 하므로 리렌더링

이렇듯 렌더링을 하는 경우가 너무나 많다 보니, **저는 state라는 파일을 따로 만들어서 보여질 restaurantList의 상태를 관리했습니다.** 이를 통해 현재 보여지는 restaurantList의 id와 내용을 따로 관리하고 이를 조작함으로써 여러 렌더링을 처리했습니다.

하지만 이 때문에 발생하는 사이드 이펙트도 분명 있었습니다. 특히 너무나도 많은 곳에서 state를 변경하고 있기에, 음식점 목록이 예상치 못한 형태로 보여졌을 때 어디서 잘못된건지 추적이 어려웠습니다.
그래서 이렇게 상태 관리를 하는 것에 대해 의문점이 들었고, 더 좋은 개선점이 있을 지 여쭤보고 싶습니다. 

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

MinSungJe committed:
fix(step1): 가게 제출 이후 값이 초기화되도록 수정

MinSungJe committed:
refactor(step1): 사용되지 않는 import 제거

MinSungJe committed:
fix(step1): 새로운 요소 추가 시 기존 요소를 삭제하지 않던 현상 수정

MinSungJe committed:
refactor(step1): RestaurantList 요소 분리

음식정 목록을 보여주는 요소를 RestaurantList.js로 분리
이를 통해 Main과 AddLunchModal 간 상호참조를 방지할 수 있음
- 음식점 목록 리렌더링 함수를 RestaurantList로 옮김, main 참조 불필요

MinSungJe committed:
refactor(step1): render 유틸 함수 및 DOM 요소 삭제

MinSungJe committed:
feat: Restaurant 타입 구현

MinSungJe committed:
feat: state.ts 구현

MinSungJe committed:
feat: Validator.ts 구현

TODO: Validator.test.js가 ts를 읽을 수 있도록 변환

MinSungJe committed:
chore|refactor: jest TypeScript 적용

MinSungJe committed:
feat: RestaurantListUtils 함수 구현

RestaurantList를 다루는 함수 모음인 RestaurantListUtils 함수 객체 구현

MinSungJe committed:
refactor: getElementList와 applyState가 State외의 다른 배열을 받을 수 있도록 변경

TODO: Default가 State로 되어 있는데, 이를 표기할 방법을 찾아야 함

MinSungJe committed:
feat: data와 state 분리

MinSungJe committed:
fix: data에 있는 정보를 잘 불러올 수 있도록 변경

MinSungJe committed:
fix: sort가 deepCopy를 하도록 변경

이로써 기존 데이터를 건드리지 않고 정렬 가능

MinSungJe committed:
feat: FilterSelect 컴포넌트 구현

MinSungJe committed:
fix: 음식점 목록이 이상한 곳에서 렌더링 되던 현상 수정

MinSungJe committed:
feat: 음식점 카테고리 정렬 기능 구현

MinSungJe committed:
feat: 음식점 거리 정렬 기능 구현

MinSungJe committed:
fix: 데이터가 변경될 때 필터의 영향을 받도록 수정

MinSungJe committed:
feat: 즐겨찾기 버튼 렌더링 구현

MinSungJe committed:
feat: 카드 별 즐겨찾기 기능 구현

MinSungJe committed:
feat: 즐겨찾기한 모든 식당을 찾는 함수 구현

MinSungJe committed:
docs: 2단계 기능요구 사항 및 프로그래밍 요구사항 작성

MinSungJe committed:
feat: 상단 nav바 디자인 추가

MinSungJe committed:
feat: 상단 nav바 클릭 시 스타일 변경 구현

MinSungJe committed:
feat: 자주 가는 음식점 nav 변경 기능 구현

MinSungJe committed:
feat: 자주 가는 음식점 표시 구현

MinSungJe committed:
feat: Card 클릭 시 모달창 열기 구현

MinSungJe committed:
feat: DetailModal 안 link 추가

MinSungJe committed:
feat: DetailModal 버튼 추가

MinSungJe committed:
fix: id 지정방식 변경

id 지정방식을 배열의 길이에서 배열 마지막 값 id에서 -1을 해서 구하도록 계산과정 변경

MinSungJe committed:
feat: DetailModal 버튼 기능 구현

닫기, 삭제하기, 즐겨찾기 버튼 기능 구현

MinSungJe committed:
fix: 이벤트 리스너를 여러 개 추가해 생긴 사이드 이펙트 제거

지금까지 모달을 연 모든 음식점이 한꺼번에 삭제되는 현상 수정

MinSungJe committed:
feat: LocalStorage 연결 구현

MinSungJe committed:
chore: 불필요한 import문 제거

MinSungJe committed:
docs: README 구현 목록 체크

MinSungJe committed:
chore: 불필요한 코드 정리

MinSungJe committed:
chore: import문 오타 수정

MinSungJe committed:
fix: DetailModal 속 즐겨찾기를 눌러 리렌더링을 할 때 조건을 부여

이를 통해 음식점 목록이 사라지는 현상을 방지할 수 있음

JeongBin0227 commented on src/component/DetailModal/DetailModalButtonContainer.js:
```diff
@@ -0,0 +1,32 @@
+        {
+          id: "close__button",
+          title: "닫기",
+          onClick: () => Modal.close("detail"),
```
detail을 넘겨주는 이유가 있을까요??

MinSungJe replied:
어떤 모달을 닫을지 id를 지정하는 역할입니다! 현재 addRestaurant 모달과 detail 모달이 있는 상태인데, 인자로 detail이라는 ID를 전달함으로써 둘 중 detail 모달에 접근할 수 있습니다.

JeongBin0227 commented on src/component/DetailModal/DetailModalButtonContainer.js:
```diff
@@ -0,0 +1,32 @@
+      TextButton.create(
+        {
+          id: "delete__button",
+          title: "삭제하기",
```
여기는 클릭이벤트가 없어도 될까요?

MinSungJe replied:
DetailModal의 버튼들을 보관하는 ButtonContainer가 아니라 Modal 안의 내용 자체를 다루는 DetailModalContent에서 클릭이벤트를 넣어주고 있었네요😢 위치 바꿔서 고치겠습니다!

JeongBin0227 commented on src/component/DetailModal/DetailModalContent.js:
```diff
@@ -0,0 +1,104 @@
+    };
+  },
+
+  handleFavoriteButton() {
+    $(".restaurant-detail .restaurant__favorite").addEventListener(
+      "click",
+      (e) => {
+        const id = e.target.id;
+        const alt = e.target.alt;
+
+        this.changeIcon(e.target);
+
+        const favoriteList = RestaurantListUtils.favoriteById(
+          LocalStorage.getJSON(RESTAURANT_LIST_KEY),
+          Number(e.target.id)
+        );
+        LocalStorage.setJSON(RESTAURANT_LIST_KEY, favoriteList);
+        state.setCurrentRestaurantList(
+          RestaurantListUtils.favoriteById(
+            state.currentRestaurantList,
+            Number(e.target.id)
+          )
+        );
+        this.renderAll();
+      }
+    );
+  },
```
여기서 favoriteById 가 두번호출되는데 한번만 호출하게 변경해볼수 있을까요??

JeongBin0227 commented on src/component/DetailModal/DetailModalContent.js:
```diff
@@ -0,0 +1,104 @@
+    return DetailModalContent;
+  },
+
+  set({ id, favorite, src, label, name, distance, description, link }) {
+    $(".restaurant-detail").innerHTML = /*html*/ `
+    <img id=${id} class="restaurant__favorite" src=${
+      favorite ? "./favorite-icon-filled.png" : "./favorite-icon-lined.png"
+    } alt=${favorite ? "favoriteIcon" : "noFavoriteIcon"} />
+    <div class="restaurant__category">
+        <img src=${src} alt=${label} />
+    </div>
+    <div class="restaurant__info">
+        <h3 class="restaurant__name text-subtitle">${name}</h3>
+        <span class="restaurant__distance text-body">캠퍼스부터 ${distance}분 내</span>
+        <p class="restaurant__description-detail text-body">${description}</p>
+        <p class="restaurant__link text-body">${link ? link : ""}</p>
+    </div>
+  `;
+
+    this.handleDeleteButton(id);
+    this.handleFavoriteButton();
+  },
```
현재 set() 함수에서 UI 업데이트(HTML 조작)와 이벤트 핸들러 등록이 함께 이루어지는거 같은데 분리해볼수 있을까요?

MinSungJe replied:
https://github.com/woowacourse/javascript-lunch/pull/253/commits/e41afc3f73b43eccae4ab3fddad76581d5af5e95 set() 함수는 데이터를 지정만하고, 핸들러 등록은 바깥에서 호출해 사용하도록 분리했습니다!

JeongBin0227 commented on src/component/FilterSelect.js:
```diff
@@ -0,0 +1,62 @@
+    return filterElement;
+  },
+
+  applyFilter(restaurantListId) {
+    const category = $("#category-filter").value;
+    const sortingRule = $("#sorting-filter").value;
+    const filteredListByCategory = this.getFilteredListByCategory(
+      LocalStorage.getJSON(RESTAURANT_LIST_KEY),
+      category
+    );
+    const filteredListByBoth = this.getFilteredListBySorting(
+      filteredListByCategory,
+      sortingRule
+    );
+    RestaurantList.applyList(restaurantListId, filteredListByBoth);
+  },
```
여기서 좀더 유연하게 categoryFilterId, sortingFilterId 를 인자로 받아서 처리해보면 어떨까요?

MinSungJe replied:
혹시 categoryFilterId와 sortingFilterId를 인자로 받아서 처리하면 왜 더 유연해지게 되는지 여쭤봐도 괜찮을까요??

기능은 어차피 고정되어 있는데,(필터링하기 + 정렬하기) 요소를 찾는 부분을 id로 받아온다고 더 유연해진다 생각이 들지 않아서요.

```js
getFilteredList(categoryFilterId, sortingFilterId) {
  const category = $(`#${categoryFilterId}`).value;
  const sortingRule = $(`#${sortingFilterId}`).value;
  const filteredListByCategory = FilterSelect.getFilteredListByCategory(
    LocalStorage.getJSON(RESTAURANT_LIST_KEY),
    category
  );
  const filteredListByBoth = FilterSelect.getFilteredListBySorting(
    filteredListByCategory,
    sortingRule
  );

  return filteredListByBoth;
},
```

JeongBin0227 replied:
밍고가 말한 것처럼 지금은 필터와 정렬의 역할이 명확하게 고정되어 있기 때문에
#category-filter, #sorting-filter를 직접 선택자로 사용해도 기능적으로 문제는 없다고 생각해용

근데 제가 categoryFilterId와 sortingFilterId를 인자로 받는 방식을 제안드린 이유는 함수의 재사용성과 유연성 때문이에요.

예를 들어, 이후에 다른 페이지에서 필터 기능을 또 쓰게 될 수도 있고, 또는 여러 개의 필터 UI를 지원하는 상황이 올 수도 있잖아요?
이럴 때 ID를 인자로 받게 되면, 하나의 getFilteredList() 함수로 여러 UI에서 필터링 기능을 공유할 수 있어요.
즉, "이 함수는 특정 DOM 구조에만 종속되지 않는다"는 걸 명시적으로 표현한 셈이에요.

당장 지금은 그 필요성이 크지 않더라도, 추후 구조가 커질 때 컴포넌트 독립성이나 테스트 용이성 측면에서도 이점이 있어서 제안드렸어요!

JeongBin0227 commented on src/component/FilterSelect.js:
```diff
@@ -0,0 +1,62 @@
+    RestaurantList.applyList(restaurantListId, filteredListByBoth);
+  },
+
+  getFilteredListByCategory(restaurantList, category) {
+    const filteredList = RestaurantListUtils.filterByCategory(
+      restaurantList,
+      category
+    );
+
+    return filteredList;
+  },
```
```
getFilteredListByCategory(restaurantList, category) {
  return RestaurantListUtils.filterByCategory(restaurantList, category);
}
```
그냥 이렇게 변경했어도 되겠네요

MinSungJe replied:
https://github.com/woowacourse/javascript-lunch/pull/253/commits/46225038f72b94ef4ac49b51602aa1ae9bc3ebde 전체적인 리팩토링을 하며 반영했습니다!

JeongBin0227 commented on src/component/RestaurantList.js:
```diff
@@ -0,0 +1,75 @@
+    return restaurantListElement;
+  },
+
+  onClickFavorite(restaurantListId, event) {
+    const target = event.target;
+    if (!target.classList.contains("restaurant__favorite")) return;
+
+    const favoriteList = RestaurantListUtils.favoriteById(
+      LocalStorage.getJSON(RESTAURANT_LIST_KEY),
+      Number(target.id)
+    );
+    LocalStorage.setJSON(RESTAURANT_LIST_KEY, favoriteList);
+    state.setCurrentRestaurantList(
+      RestaurantListUtils.favoriteById(
+        state.currentRestaurantList,
+        Number(target.id)
+      )
+    );
+    this.applyState(restaurantListId);
+  },
```
여기도 불필요한 상태 업데이트 제거하고, favoriteById()를 한 번만 호출하여 최적화해볼수 있지 않을까요?

JeongBin0227 commented on src/component/RestaurantList.js:
```diff
@@ -0,0 +1,75 @@
+  },
+};
+
+export default RestaurantList;
```
applyList와 applyState 는 각각 어떤역할을 하나요?

MinSungJe replied:
applyList는 해당 RestaurantList Element에 List로 전달한 음식점 목록을 반영하는 함수입니다.
applyState는 해당 RestaurantList Element에 전역으로 관리되는 음식점 목록 상태를 반영하는 함수입니다.

https://github.com/woowacourse/javascript-lunch/pull/253/commits/0cc79c4ec48ad9c76e43837471fa5d44c59e08b8, https://github.com/woowacourse/javascript-lunch/pull/253/commits/a424d885825c0786992f494b84af6933e9f61ad8 생각해보니, applyList 외의 나머지 함수들은 필요가 없어보여, 전체적으로 리팩토링을 진행했습니다! 😊 상세한 변경점은 코멘트에 적어두겠습니다.

JeongBin0227 commented on src/component/RestaurantList.js:
```diff
@@ -0,0 +1,75 @@
+    this.applyElements(restaurantListId, restaurantElementList);
+  },
+
+  applyElements(restaurantListId, elements) {
+    $(`.restaurant-list[id=${restaurantListId}]`).replaceChildren();
+    elements.forEach((element) =>
+      $(`.restaurant-list[id=${restaurantListId}]`).appendChild(element)
+    );
+  },
```
매번 querySelector()를 호출하여 성능 저하가 발생할 수 도있는데, document.getElementById()를 사용하여 한 번만 요소를 찾고, innerHTML = ""로 기존 요소 삭제 후 append() 사용해보는 방식과 어떤 차이가 있을까요??

MinSungJe replied:
위에서도 그렇고 전체적으로 두 번씩 `querySelector()`를 선언하는 경우가 많았네요😢
`replaceChildren()`과 `innerHTML = ""`은 기능에 차이가 없어보이고, querySelector를 두 번 부르는 부분만 수정했습니다! https://github.com/woowacourse/javascript-lunch/pull/253/commits/d98f8ae19122bcd9bfde82ae13db672b86695e9d

JeongBin0227 commented on src/utils/RestaurantListUtils.ts:
```diff
@@ -0,0 +1,54 @@
+  },
+};
+
+export default RestaurantListUtils;
```
sortById(), sortByName(), sortByDistance()에서 불필요한 배열 복사하는거 같아요. 이를 개선해보면 어떨까요?

MinSungJe replied:
`Array.sort()`가 기존의 Array를 정렬해버려서 복사해서 붙여넣었는데,(원본의 불변성을 해쳐서) 생각해보니 함수에 전달할 때 복사를 먼저하네요! 복사없이 배열을 반환하도록 개선했습니다. https://github.com/woowacourse/javascript-lunch/pull/253/commits/03010ac29e25c0ef6540b723651c903e81445fa6

```js
getFilteredListBySorting(restaurantList, sortingRule) {
  // 미리 복사 진행후 전달
  let filteredList = [...restaurantList];
  if (sortingRule === "id")
    filteredList = RestaurantListUtils.sortById(filteredList);
  if (sortingRule === "name")
    filteredList = RestaurantListUtils.sortByName(filteredList);
  if (sortingRule === "distance")
    filteredList = RestaurantListUtils.sortByDistance(filteredList);

  return filteredList;
},
```

JeongBin0227 commented on src/utils/RestaurantListUtils.ts:
```diff
@@ -0,0 +1,54 @@
+    return resultList.sort((a, b) => a.distance - b.distance);
+  },
+
+  favoriteById(restaurantList: Restaurant[], id: number) {
+    return restaurantList.map((restaurant) =>
+      restaurant.id === id
+        ? { ...restaurant, favorite: !restaurant.favorite }
+        : restaurant
+    );
+  },
```
map()에서 Object.assign()을 활용하면 어떤 효과가 있을까요?

MinSungJe replied:
Spread Operator를 이용하면 무조건 새로운 객체를 만들게 되는데 반해, `Object.assign()`을 이용하면 객체의 값을 바로 붙여 넣을 수 있어 불필요한 복사를 줄이는 효과가 있겠네요, 하지만 원본 배열을 해칠 수 있어 map함수를 이용해 새 배열을 반환하는 이유가 없어질 것 같습니다.

JeongBin0227 commented:
> 이외에도 TypeScript를 익히기 위해 UI가 아닌 코드와 이에 대한 단위 테스트 코드를 전부 TS로 변환해 작성했습니다

-> 밍고가 생각하는 TS 의 장점은 무엇일까요?

JeongBin0227 commented:
> applyFilter의 위치

-> RestaurantList는 주어진 리스트를 출력하는 책임만 져야 한다고 생각합니다.
즉,  applyFilter()의 역할과 책임 분리를 고민하기 전에

RestaurantList → 리스트를 받아서 UI를 렌더링하는 역할 (책임: "표시"만 담당)
FilterSelect → 사용자가 선택한 필터를 적용하여 새로운 리스트를 만드는 역할 (책임: "데이터 가공")
라고 생각해요

따라서 applyFilter()는 필터링된 데이터를 만들어야 하므로 FilterSelect에 두는 것이 맞다고 생각합니다.

이제 이 전제를 두고 밍고의 고민을 생각해보면

FilterSelect에서 필터링된 데이터만 반환하고, 상태 관리 로직에서 적용해보면 어떨까요? 좀더 풀어서 생각해보면 FilterSelect는 단순히 필터링된 데이터를 제공하는 역할만 수행하고 상태에서 필터링된 데이터를 가져와 적용합니다. 그러면 UI에서 state.currentRestaurantList를 자동으로 반영하여 렌더링을 자연스럽게 연결됩니다.
이렇게하면 applyFilter()를 직접 UI에서 호출할 필요 없이, 상태 관리에서 알아서 처리되도록 변경 가능합니다.

혹시나 이해가 안된다면 다시 질문주세요!

JeongBin0227 reviewed (changes requested):
안녕하세요~ 밍고!

2단계 코드 잘 해주셨습니다 👍 하지만 코드 변경점이 많다보니 제가 따라가기가 너무 어려웠네요 ㅠㅠ 우선 1차로 코드 전반적인 부분이랑 질문주신 부분 위주로 코멘트 달아두었습니다!
이해가 안되는 코멘트 있으면 말씀주세요!

미션하느라 고생많으셨어요~

JeongBin0227 commented:
> 이번 점심 뭐먹지는 렌더링과의 싸움이었습니다.
> 
> 음식점 목록에서 즐겨찾기를 누르면 아이콘 변화를 위해 음식점 목록을 리렌더링해야 됨
> DetailModal에서 즐겨찾기를 눌렀을 때 바깥 컴포넌트에도 적용되야 하므로 리렌더링
> 삭제하기 버튼을 눌렀을 때도 리렌더링
> 필터를 고를때도 맞춰 목록이 표시되야 하므로 리렌더링
> 이렇듯 렌더링을 하는 경우가 너무나 많다 보니, 저는 state라는 파일을 따로 만들어서 보여질 restaurantList의 상태를 관리했습니다. 이를 통해 현재 보여지는 restaurantList의 id와 내용을 따로 관리하고 이를 조작함으로써 여러 렌더링을 처리했습니다.
> 
> 하지만 이 때문에 발생하는 사이드 이펙트도 분명 있었습니다. 특히 너무나도 많은 곳에서 state를 변경하고 있기에, 음식점 목록이 예상치 못한 형태로 보여졌을 때 어디서 잘못된건지 추적이 어려웠습니다.
> 그래서 이렇게 상태 관리를 하는 것에 대해 의문점이 들었고, 더 좋은 개선점이 있을 지 여쭤보고 싶습니다.

-> 우선 바닐라 js 에서 어떻게 상태를 관리할수 있는지 한번 공부해보면 좋을거 같아요! 그러면 dataset 속성을 활용한 상태 변경 감지하거나, EventEmitter 패턴을 이용하거나.. 많은 방법이 나오는데 그중에서 밍고의 구조에서 어떤게 제일 적절할지 고민해보면 좋을거같아요!
말씀하신것처럼 바닐라 js에서 제일 어려운게 렌더링이라고 생각해요. 밍고는 이미 리액트를 사용해보셨을지 모르겠지만, 리액트에서는 이런것들을 고민안해도 되어서 지금 고민해보는게 너무 적절할거 같아서 정답을 알려주기 보다는 물음을 남길게용 혹시나 어려우면 언제든지 말씀해주세요!

MinSungJe committed:
refactor: Renderer 분리 및 DetailModalButtonContainer로 onClick 이동

MinSungJe committed:
refactor: FilterSelect와 렌더 책임 분리

Renderer에 렌더 책임을 넣고 FilterSelect에는 Filter된 리스트만 나오도록 분리

MinSungJe committed:
chore: Renderer.renderOuter 함수명 변경

MinSungJe committed:
refactor: applyData, applyState 제거

Renderer 분리로 필요없어진 applyData, applyState 제거

MinSungJe committed:
fix: 자주 가는 음식점에서 즐겨찾기가 되지 않던 현상 수정

main의 initNavigationButton() 중 눌렀을 때 현재 탭 상황 반영코드가 누락되어 있었음
정상 반영 및 해당 함수 코드 정리

MinSungJe committed:
refactor: DetailModalContent UI 업데이트와 이벤트 핸들러 등록 분리

MinSungJe committed:
chore: Renderer.filteredList 함수명 변경

MinSungJe committed:
fix: 자주 가는 음식점 탭을 보고 다시 모든 음식점 탭을 표시할 때 필터가 사라지는 현상 수정

MinSungJe committed:
refactor: applyElements 내 중복되는 querySelector 제거

MinSungJe committed:
refactor: 불필요한 배열 복사 제거

MinSungJe committed:
refactor: state 제거

MinSungJe committed:
chore: 불필요한 import문 제거

MinSungJe commented:
안녕하세요 케빈! 정말 늦었지만 2단계 리뷰에 대한 첫 번째 요청 드립니다.
코드 변경점이 많다보니 따라가기가 너무 어려웠단 말씀을 듣고..😢 어떻게 하면 따라가기 쉬운 코드를 작성할 수 있을까 생각하며 코드를 수정했습니다.

가장 큰 문제는 렌더링이 이곳저곳에서 되니, 이를 따라가기 힘들 것 같다는 생각이 들었습니다. 그래서 **렌더링을 하는 부분을 전부 찾아서 Renderer라는 함수 객체 안의 기능으로 대체**했습니다. 그렇게 대체하니 favoriteList를 두 번 querySelector로 찾는 문제도 자연스레 해결됐습니다. https://github.com/woowacourse/javascript-lunch/pull/253/commits/88d2022c21438cec3578c6f26ca84c0514395eaf https://github.com/woowacourse/javascript-lunch/pull/253/commits/46225038f72b94ef4ac49b51602aa1ae9bc3ebde



> -> 밍고가 생각하는 TS 의 장점은 무엇일까요?

제가 생각하는 TS의 장점은 변수에 들어갈 값을 예상할 수 있게 된다는 점입니다. 미리 TS로 정의한 변수를 사용하면 vscode의 자동완성 기능으로 사용할 수 있는 프로퍼티나 메서드를 추천받을 수 있습니다. 따라서 제가 불러온 변수가 의도랑 맞는지 점검할 수 있고, 타입을 잘 모르겠는 경우도 일단 적고보면 어떤 메서드가 있는지 확인하여 무슨 종류인지 알 수 있습니다.

이러한 편리성 및 타입오류에 대한 예방이 제가 생각한 TS의 가장 큰 장점입니다.

> 우선 바닐라 js 에서 어떻게 상태를 관리할수 있는지 한번 공부해보면 좋을거 같아요! 그러면 dataset 속성을 활용한 상태 변경 감지하거나, EventEmitter 패턴을 이용하거나.. 많은 방법이 나오는데 그중에서 밍고의 구조에서 어떤게 제일 적절할지 고민해보면 좋을거같아요!

케빈이 남겨준 코멘트 덕분에 바닐라 JS에서 상태를 관리하는 법을 많이 공부해봤습니다. 대부분의 경우가 Observer 패턴으로, 상태가 바뀌었을 때 구독하고 있던 요소들에게 `나 바뀌었어!`하고 티를 내는 구조더라구요. 

- 정보 변경 요청 -> state 변경 -> state가 변경되었다고 구독한 요소에게 알림 -> 알림받은 요소는 state 정보로 재렌더링

저의 경우엔 DOM 속 정보에 먼저 접근해 값을 바꾸고 state를 같이 바꾸는 형태이지만 이러면 state로 요소를 변경한다기보단, 현재 상태가 어떤지 기록만 하는 정도에만 그치지 않는다고 느꼈습니다.

- 정보 변경 요청 -> 정보 변경 및 state도 맞춰서 변경

따라서 저는 **기록만 하고 있을 뿐 실제로 반영하지않는 잘못된 방식으로 상태를 사용하고 있다 판단, 관련 코드를 전부 제거**하여 상태 저장 없이 조작에 맞춰 렌더링만 하도록 리팩토링을 진행하였습니다. https://github.com/woowacourse/javascript-lunch/pull/253/commits/a424d885825c0786992f494b84af6933e9f61ad8

이 때 현재 열려있는 탭 상태를 저장하는 상태는 dataset으로 값을 저장하는 것처럼 탭 버튼에 activated라는 class가 있는지에 따라 구분하도록 대체하였습니다.

케빈의 질문 덕분에 앞으로의 미션에서 상태를 고려할 땐, **상태가 변경됨에 따라 자동으로 상태가 반영이 되는가?에 주목**해야겠다는 생각이 들었습니다. 덕분에 상태의 조건에 대해 확고해진 것 같아 뿌듯하네요😊

JeongBin0227 reviewed (approved):
안녕하세요 밍고!

코멘트가 좀 많을수도 있었는데, 잘 반영해주셔서 감사합니다! 수정해주신 부분들 확인했고 충분히 머지할만한 코드라고 생각해서 여기서 머지하도록 하겠습니다

점심 뭐먹지 미션 완료하신거 축하드려요~ 다음미션도 화이팅입니다 💯
